File: apps/server/src/traversal/deadlines.ts
// Pure date math for the traversal engine. No I/O — unit tested directly.

export interface DeadlineInputs {
  since: string; // ISO date "YYYY-MM-DD", when the license was first held
  lastFiledAt: string | null; // last filing date for this requirement, if any
  intervalMonths: number | null; // null = no recurring cadence
  dueMonthDay: string | null; // "MM-DD", snaps the computed date to a fixed calendar day
  graceDays: number | null;
}

export type ObligationStatus = "overdue" | "due_soon" | "upcoming" | "no_deadline";

export interface DeadlineResult {
  nextDueDate: string | null; // ISO date "YYYY-MM-DD"
  status: ObligationStatus;
}

const DUE_SOON_WINDOW_DAYS = 30;
const MS_PER_DAY = 24 * 60 * 60 * 1000;

/** Parse "YYYY-MM-DD" as a local-midnight Date, avoiding UTC-parsing offset bugs. */
function parseISODate(iso: string): Date {
  const [y, m, d] = iso.split("-").map(Number);
  return new Date(y, m - 1, d);
}

function formatISODate(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

/** Truncate a Date to local midnight, for day-granularity comparisons. */
function atMidnight(date: Date): Date {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
}

/**
 * Add months to a date, clamping to the last day of the target month when the
 * source day-of-month doesn't exist there (e.g. Jan 31 + 1 month -> Feb 28, not
 * a rollover into March). This is the "month-end rollover" behavior under test.
 */
export function addMonths(date: Date, months: number): Date {
  const targetMonthIndex = date.getMonth() + months;
  const daysInTargetMonth = new Date(date.getFullYear(), targetMonthIndex + 1, 0).getDate();
  const day = Math.min(date.getDate(), daysInTargetMonth);
  return new Date(date.getFullYear(), targetMonthIndex, day);
}

/** Snap a date's month/day to a fixed "MM-DD", keeping the date's year. */
function snapToMonthDay(date: Date, monthDay: string): Date {
  const [month, day] = monthDay.split("-").map(Number);
  const daysInMonth = new Date(date.getFullYear(), month, 0).getDate();
  return new Date(date.getFullYear(), month - 1, Math.min(day, daysInMonth));
}

function computeStatus(nextDue: Date, graceDays: number, now: Date): ObligationStatus {
  const diffDays = Math.round((nextDue.getTime() - atMidnight(now).getTime()) / MS_PER_DAY);
  if (diffDays < -graceDays) return "overdue";
  if (diffDays <= DUE_SOON_WINDOW_DAYS) return "due_soon";
  return "upcoming";
}

export function computeDeadline(inputs: DeadlineInputs, now: Date = new Date()): DeadlineResult {
  if (inputs.intervalMonths === null) {
    return { nextDueDate: null, status: "no_deadline" };
  }

  const base = parseISODate(inputs.lastFiledAt ?? inputs.since);
  let nextDue = addMonths(base, inputs.intervalMonths);
  if (inputs.dueMonthDay) {
    nextDue = snapToMonthDay(nextDue, inputs.dueMonthDay);
  }

  const graceDays = inputs.graceDays ?? 0;
  return {
    nextDueDate: formatISODate(nextDue),
    status: computeStatus(nextDue, graceDays, now),
  };
}

File: apps/engine/payna_engine/graph/deadlines.py
"""Pure date math for filing deadlines — a direct port of
apps/server/src/traversal/deadlines.ts, including month-end rollover clamping.
"""

from __future__ import annotations

import calendar
from dataclasses import dataclass
from datetime import date
from typing import Literal, Optional

ObligationStatus = Literal["overdue", "due_soon", "upcoming", "no_deadline"]

DUE_SOON_WINDOW_DAYS = 30


@dataclass
class DeadlineInputs:
    since: str
    last_filed_at: Optional[str]
    interval_months: Optional[int]
    due_month_day: Optional[str]
    grace_days: Optional[int]


@dataclass
class DeadlineResult:
    next_due_date: Optional[str]
    status: ObligationStatus


def _parse_iso(iso: str) -> date:
    y, m, d = (int(p) for p in iso.split("-"))
    return date(y, m, d)


def add_months(d: date, months: int) -> date:
    """Add months, clamping to the last valid day of the target month."""
    total = d.month - 1 + months
    year = d.year + total // 12
    month = total % 12 + 1
    days_in_month = calendar.monthrange(year, month)[1]
    return date(year, month, min(d.day, days_in_month))


def _snap_to_month_day(d: date, month_day: str) -> date:
    month, day = (int(p) for p in month_day.split("-"))
    days_in_month = calendar.monthrange(d.year, month)[1]
    return date(d.year, month, min(day, days_in_month))


def _status(next_due: date, grace_days: int, now: date) -> ObligationStatus:
    diff_days = (next_due - now).days
    if diff_days < -grace_days:
        return "overdue"
    if diff_days <= DUE_SOON_WINDOW_DAYS:
        return "due_soon"
    return "upcoming"


def compute_deadline(inputs: DeadlineInputs, now: Optional[date] = None) -> DeadlineResult:
    now = now or date.today()
    if inputs.interval_months is None:
        return DeadlineResult(next_due_date=None, status="no_deadline")

    base = _parse_iso(inputs.last_filed_at or inputs.since)
    next_due = add_months(base, inputs.interval_months)
    if inputs.due_month_day:
        next_due = _snap_to_month_day(next_due, inputs.due_month_day)

    grace = inputs.grace_days or 0
    return DeadlineResult(next_due_date=next_due.isoformat(), status=_status(next_due, grace, now))
